### PR TITLE
enhancement: allow toggling custom config deprecation behaviour

### PR DESCRIPTION
**Pull request by rexagod**
## Description

Allow toggling custom config deprecation behaviour; helpful in scenarios where a controller disables all service discovery methods.

## Type of change

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [ ] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [x] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Changelog entry

```release-note
Users can toggle Prometheus controller's custom configuration deprecation behavior using the "--deprecate-custom-configuration" flag.
```

### CODE REVIEW

File: Documentation/operator.md [hunk 1]
```diff
@@ -49,6 +49,8 @@ Usage of ./operator:
     	Value used by the operator to filter Alertmanager, Prometheus, PrometheusAgent and ThanosRuler objects that it should reconcile. If the value isn't empty, the operator only reconciles objects with an operator.prometheus.io/controller-id annotation of the same value. Otherwise the operator reconciles all objects without the annotation or with an empty annotation value.
   -deny-namespaces value
     	Namespaces not to scope the interaction of the Prometheus Operator (deny list). This is mutually exclusive with --namespaces.
+  -deprecate-custom-configuration
```
Comment (simonpasquier) [inline Documentation/operator.md:52]: I'd suggest to name the flag `-disable-unmanaged-prometheus-configuration` and default to false.

**Comment by mviswanathsai** [comment]
Not sure I understand what the flag is for, can you elaborate a bit? I lack the context behind this.

**Comment by simonpasquier** [comment]
To provide more context, there's a legacy behavior which instructs the operator to not manage the Prometheus configuration when all selectors (ruleSelector, serviceMonitorSelector, ...) are nil. In this case, the user is supposed to build by hand the secret holding the (gzipped) configuration. It can be a pain to debug if you run into this mode by accident because the operator doesn't complain but the Prometheus statefulset will never spin up pods if the secret is missing.
This is very much deprecated now since we have additionalScrapeConfigs + ScrapeConfig CRD but removing it would break users. Adding a flag to say "don't support unamanaged configuration mode" improves the UX.

**Comment by rexagod** [comment]
This PR is rebased on https://github.com/prometheus-operator/prometheus-operator/pull/6992 to address an issue that crashed the `make generate` target.

File: cmd/operator/main.go [hunk 2]
```diff
@@ -183,7 +185,7 @@ func parseFlags(fs *flag.FlagSet) {
 	fs.Var(&cfg.SecretListWatchLabelSelector, "secret-label-selector", "Label selector to filter Secrets to watch")
 
 	fs.Float64Var(&memlimitRatio, "auto-gomemlimit-ratio", defaultMemlimitRatio, "The ratio of reserved GOMEMLIMIT memory to the detected maximum container or system memory. The value should be greater than 0.0 and less than 1.0. Default: 0.0 (disabled).")
-
+	fs.BoolVar(&disableUnmanagedPrometheusConfiguration, "disable-unmanaged-prometheus-configuration", false, "Toggle the legacy deprecation behaviour for unmanaged Prometheus configuration. Default: false.")
```
Comment (simonpasquier) [inline cmd/operator/main.go:188]:
```suggestion
	fs.BoolVar(&disableUnmanagedPrometheusConfiguration, "disable-unmanaged-prometheus-configuration", false, "Disable support for unmanaged Prometheus configuration when all resource selectors are nil. Default: false.")
```

File: pkg/prometheus/server/operator.go [hunk 3]
```diff
@@ -117,6 +119,14 @@ func WithStorageClassValidation() ControllerOption {
 	}
 }
 
+// WithDisableUnmanagedPrometheusConfiguration tells that the controller should deprecate
+// the custom configuration.
+func WithDisableUnmanagedPrometheusConfiguration() ControllerOption {
```
Comment (simonpasquier) [inline pkg/prometheus/server/operator.go:124]:
```suggestion
func WithUnmanagedConfigurationDisabled() ControllerOption {
```

File: pkg/prometheus/server/operator.go [hunk 4]
```diff
@@ -91,6 +91,8 @@ type Operator struct {
 	canReadStorageClass    bool
 
 	eventRecorder record.EventRecorder
+
+	disableUnmanagedPrometheusConfiguration bool
```
Comment (simonpasquier) [inline pkg/prometheus/server/operator.go:95]:
```suggestion
	disableUnmanagedConfiguration bool
```

File: pkg/prometheus/server/operator.go [hunk 5]
```diff
@@ -954,7 +964,7 @@ func (c *Operator) UpdateStatus(ctx context.Context, key string) error {
 	return nil
 }
 
-func logDeprecatedFields(logger *slog.Logger, p *monitoringv1.Prometheus) {
+func logDeprecatedFields(logger *slog.Logger, p *monitoringv1.Prometheus, disableUnmanagedPrometheusConfiguration bool) {
```
Comment (simonpasquier) [inline pkg/prometheus/server/operator.go:967]: (suggestion) let's make it a pointer receiver function.
```suggestion
func (c *Operator) logDeprecatedFields(logger *slog.Logger, p *monitoringv1.Prometheus) {
```

File: cmd/operator/main.go [hunk 6]
```diff
@@ -274,6 +276,9 @@ func run(fs *flag.FlagSet) int {
 		promControllerOptions         = []prometheuscontroller.ControllerOption{}
 		thanosControllerOptions       = []thanoscontroller.ControllerOption{}
 	)
+	if disableUnmanagedPrometheusConfiguration {
+		promControllerOptions = append(promControllerOptions, prometheuscontroller.WithDisableUnmanagedPrometheusConfiguration())
```
Comment (simonpasquier) [inline cmd/operator/main.go:280]: it might be good to log an info message saying that unmanaged configuration is disabled.

**Review by simonpasquier — COMMENTED** [review]
As a follow-up change, I suggest to enhance the `Reconciled` status condition so that when all selectors are nil, it gets reflected in the reconciled message. E.g. it's not an error but it can be hard to find out what you missed. Something like:


````
status:
  availableReplicas: 2
  conditions:
  - lastTransitionTime: "2024-10-21T14:45:40Z"
    message: ""
    observedGeneration: 2
    reason: ""
    status: "True"
    type: Available
  - lastTransitionTime: "2024-10-21T14:45:40Z"
    message: "None of the scrape resource selectors are specified."
    observedGeneration: 2
    reason: "NoScrapeResourceSelected"
    status: "True"
    type: Reconciled

````

File: Documentation/operator.md [hunk 7]
```diff
@@ -49,6 +49,8 @@ Usage of ./operator:
     	Value used by the operator to filter Alertmanager, Prometheus, PrometheusAgent and ThanosRuler objects that it should reconcile. If the value isn't empty, the operator only reconciles objects with an operator.prometheus.io/controller-id annotation of the same value. Otherwise the operator reconciles all objects without the annotation or with an empty annotation value.
   -deny-namespaces value
     	Namespaces not to scope the interaction of the Prometheus Operator (deny list). This is mutually exclusive with --namespaces.
+  -disable-unmanaged-prometheus-configuration
+    	Disable support for unmanaged Prometheus configuration when all resource selectors are nil. Default: false.
```
Comment (simonpasquier) [inline Documentation/operator.md:53]: (nit) let's provide more context why you might want to use the flag.
```suggestion
    	Disable support for unmanaged Prometheus configuration when all resource selectors are nil. As stated in the API documentation, unmanaged Prometheus configuration is a deprecated feature which can be avoided with "additionalScrapeConfigs` or the ScrapeConfig CRD. Default: false.
```

File: pkg/prometheus/server/operator.go [hunk 8]
```diff
@@ -91,6 +91,8 @@ type Operator struct {
 	canReadStorageClass    bool
 
 	eventRecorder record.EventRecorder
+
+	disableUnmanagedConfiguration bool
```
Comment (simonpasquier) [inline pkg/prometheus/server/operator.go:95]: (nit) can we move it next to the other bool fields (e.g. at L92)?

File: pkg/prometheus/server/operator.go [hunk 9]
```diff
@@ -117,6 +119,14 @@ func WithStorageClassValidation() ControllerOption {
 	}
 }
 
+// WithUnmanagedConfigurationDisabled tells that the controller should deprecate
+// the custom configuration.
+func WithUnmanagedConfigurationDisabled() ControllerOption {
```
Comment (simonpasquier) [inline pkg/prometheus/server/operator.go:124]: (nit)

```suggestion
func WithoutUnmanagedConfiguration() ControllerOption {
```

**Review by simonpasquier — COMMENTED** [review]
I've got just a few last nits otherwise LGTM.

File: Documentation/operator.md [hunk 10]
```diff
@@ -49,6 +49,8 @@ Usage of ./operator:
     	Value used by the operator to filter Alertmanager, Prometheus, PrometheusAgent and ThanosRuler objects that it should reconcile. If the value isn't empty, the operator only reconciles objects with an operator.prometheus.io/controller-id annotation of the same value. Otherwise the operator reconciles all objects without the annotation or with an empty annotation value.
   -deny-namespaces value
     	Namespaces not to scope the interaction of the Prometheus Operator (deny list). This is mutually exclusive with --namespaces.
+  -disable-unmanaged-prometheus-configuration additionalScrapeConfigs
```
Comment (simonpasquier) [inline Documentation/operator.md:52]: hmm, not sure why we have `additionalScrapeConfigs` here but it looks like removing the backticks in the help text should fix it :exploding_head:

File: cmd/operator/main.go [hunk 11]
```diff
@@ -183,7 +185,7 @@ func parseFlags(fs *flag.FlagSet) {
 	fs.Var(&cfg.SecretListWatchLabelSelector, "secret-label-selector", "Label selector to filter Secrets to watch")
 
 	fs.Float64Var(&memlimitRatio, "auto-gomemlimit-ratio", defaultMemlimitRatio, "The ratio of reserved GOMEMLIMIT memory to the detected maximum container or system memory. The value should be greater than 0.0 and less than 1.0. Default: 0.0 (disabled).")
-
+	fs.BoolVar(&disableUnmanagedPrometheusConfiguration, "disable-unmanaged-prometheus-configuration", false, "Disable support for unmanaged Prometheus configuration when all resource selectors are nil. As stated in the API documentation, unmanaged Prometheus configuration is a deprecated feature which can be avoided with `additionalScrapeConfigs` or the ScrapeConfig CRD. Default: false.")
```
Comment (simonpasquier) [inline cmd/operator/main.go:188]:
```suggestion
	fs.BoolVar(&disableUnmanagedPrometheusConfiguration, "disable-unmanaged-prometheus-configuration", false, "Disable support for unmanaged Prometheus configuration when all resource selectors are nil. As stated in the API documentation, unmanaged Prometheus configuration is a deprecated feature which can be avoided with '.spec.additionalScrapeConfigs' or the ScrapeConfig CRD. Default: false.")
```

File: cmd/operator/main.go [hunk 12]
```diff
@@ -274,6 +276,10 @@ func run(fs *flag.FlagSet) int {
 		promControllerOptions         = []prometheuscontroller.ControllerOption{}
 		thanosControllerOptions       = []thanoscontroller.ControllerOption{}
 	)
+	if disableUnmanagedPrometheusConfiguration {
+		logger.Info("unmanaged Prometheus configuration is disabled")
```
Comment (simonpasquier) [inline cmd/operator/main.go:280]:
```suggestion
		logger.Info("Disabling support for unmanaged Prometheus configurations")
```

File: pkg/prometheus/server/operator.go [hunk 13]
```diff
@@ -117,6 +118,14 @@ func WithStorageClassValidation() ControllerOption {
 	}
 }
 
+// WithoutUnmanagedConfiguration tells that the controller should deprecate
+// the custom configuration.
```
Comment (simonpasquier) [inline pkg/prometheus/server/operator.go:122]:
```suggestion
// WithoutUnmanagedConfiguration tells that the controller should not support
// unmanaged configurations.
```

**Review by simonpasquier — COMMENTED** [review]
A few nits but lgtm

**Review by simonpasquier — APPROVED** [review]

**Comment by simonpasquier** [comment]
Thanks!

**Pull request merged**